### PR TITLE
Fix `OverflowError` when SSL is used on some Python versions

### LINKED ISSUE #2513: OverflowError: signed integer is greater than maximum
Both of the below calls result in an `OverflowError` being raised because the `Content-Length` is longer than 3GB. OpenSSL only supports reading (2**31)-1 bytes in a single read call.

```python
url = "https://dailymed-data.nlm.nih.gov/public-release-files/dm_spl_release_human_rx_part1.zip"

resp = http.request("GET", url)
resp.data

resp = http.request("GET", url, preload_content=False)
list(resp.stream(None))
```

```
Traceback (most recent call last):
  File "/home/sethmlarson/urllib3/testme.py", line 5, in <module>
    resp = http.request("GET", "https://dailymed-data.nlm.nih.gov/public-release-files/dm_spl_release_human_rx_part1.zip")
  File "/home/sethmlarson/urllib3/src/urllib3/__init__.py", line 112, in request
    return _DEFAULT_POOL.request(
  File "/home/sethmlarson/urllib3/src/urllib3/_request_methods.py", line 109, in request
    return self.request_encode_url(
  File "/home/sethmlarson/urllib3/src/urllib3/_request_methods.py", line 142, in request_encode_url
    return self.urlopen(method, url, **extra_kw)
  File "/home/sethmlarson/urllib3/src/urllib3/poolmanager.py", line 443, in urlopen
    response = conn.urlopen(method, u.request_uri, **kw)
  File "/home/sethmlarson/urllib3/src/urllib3/connectionpool.py", line 744, in urlopen
    response = self.ResponseCls.from_httplib(
  File "/home/sethmlarson/urllib3/src/urllib3/response.py", line 748, in from_httplib
    resp = ResponseCls(
  File "/home/sethmlarson/urllib3/src/urllib3/response.py", line 467, in __init__
    self._body = self.read(decode_content=decode_content)
  File "/home/sethmlarson/urllib3/src/urllib3/response.py", line 664, in read
    data = self._fp.read() if not fp_closed else b""
  File "/home/sethmlarson/.pyenv/versions/3.8.6/lib/python3.8/http/client.py", line 471, in read
    s = self._safe_read(self.length)
  File "/home/sethmlarson/.pyenv/versions/3.8.6/lib/python3.8/http/client.py", line 612, in _safe_read
    data = self.fp.read(amt)
  File "/home/sethmlarson/.pyenv/versions/3.8.6/lib/python3.8/socket.py", line 669, in readinto
    return self._sock.recv_into(b)
  File "/home/sethmlarson/.pyenv/versions/3.8.6/lib/python3.8/ssl.py", line 1241, in recv_into
    return self.read(nbytes, buffer)
  File "/home/sethmlarson/.pyenv/versions/3.8.6/lib/python3.8/ssl.py", line 1099, in read
    return self._sslobj.read(len, buffer)
OverflowError: signed integer is greater than maximum
```

The fix would likely need to go into `urllib3.response.HTTPResponse` to limit the amount of data requested in a single call to `http.client.HTTPResponse.read(amt)`.

Originally reported in https://github.com/psf/requests/issues/5969

### Minimum requirements

:dollar: **You can get paid to complete this issue! [Please read the docs for more information](https://urllib3.readthedocs.io/en/latest/contributing.html#getting-paid-for-your-contributions).**

- [ ] Ensure that any call to `urllib3.response.HTTPResponse.read(...)` never results in a call to `.read(n)` with a value greater than `n=(2**31)-1`. Behavior should not change for users, you should buffer bytes into a `bytearray` and then return the stitched together data when necessary.
- [ ] Test cases for streaming chunk sizes, `.read(2**31)`, `.read(None)`, `.read()`, `preload_content=True`
- [ ] Merge fix into both `main` and `1.26.x` branches

### PR DESCRIPTION
**Pull request by illia-v**
Fixes #2513.

### CODE REVIEW

File: src/urllib3/response.py
Comment (pquentin) [inline src/urllib3/response.py:715]: Let's not consider unreleased versions of Python please :)

File: src/urllib3/response.py
Comment (pquentin) [inline src/urllib3/response.py:739]: Can we please update this comment? `cStringIO` was removed in Python 3. (Can be done in a future PR)

File: src/urllib3/response.py
Comment (pquentin) [inline src/urllib3/response.py:721]: Since we only use `buffer.write()` and `buffer.getvalue()`, why not use a simple `bytearray`?

File: src/urllib3/response.py
Comment (pquentin) [inline src/urllib3/response.py:718]: While nested functions are nice, they're slower than normal functions since the function object gets built for every invocation and can be confusing. (Generally speaking, pretending Python is a functional language never ends well. Tail recursion is the other good example.)

Can you please:
 * place this nested function at the same level as the existing `read()` function? Maybe call it `_read_aux` or `_read_chunks`
 * place the `if` *inside* this new function?

File: src/urllib3/response.py
Comment (pquentin) [inline src/urllib3/response.py:732]: OK so we're talking about gigabytes of memory here. For a 3GB read, what is the peak memory usage? Is there a way to ensure we never use more than 3GB of memory?

**Review by pquentin — CHANGES_REQUESTED** [review]
Thanks! The approach looks good.

File: src/urllib3/response.py
Comment (jayaddison) [reply src/urllib3/response.py:715]: Relocating a comment from https://github.com/illia-v/urllib3/commit/2d6cf6c6f2714e5ad6455c1e665af83e9af9293f#r76646438 (thank you @illia-v for redirecting here):

Would it be worth considering using [`util.ssl_._is_ge_openssl_v1_1_1`](https://github.com/urllib3/urllib3/blob/f9c7434090ce96ca93180137356531a9cfcce822/src/urllib3/util/ssl_.py#L28-L38) -- or something like it -- to check whether to enable this logic?

One of my thoughts was that that could allow people who compile `cpython` themselves (for example, with different SSL libraries/version) and who use other variants of Python to benefit from the fix.

Comment (pquentin) [reply src/urllib3/response.py:715]: Sure, if this fix depends on the OpenSSL version, we should definitely check that instead. I haven't investigated enough to tell.

Comment (illia-v) [reply src/urllib3/response.py:715]: @jayaddison I can reproduce the overflow on CPython 3.8 compiled against OpenSSL 1.0.2 and 1.1.1. Does it really depend on the OpenSSL version?

Comment (sethmlarson) [reply src/urllib3/response.py:715]: I was also under the impression that this was an issue in Python, not with OpenSSL.

File: src/urllib3/response.py
Comment (sethmlarson) [reply src/urllib3/response.py:739]: Let's remove now since we're touching this area and the comment is out of date.

File: src/urllib3/response.py
Comment (jayaddison) [reply src/urllib3/response.py:715]: Hmm, yep, I think I misunderstood something.  Re-constructing the situation:

The underlying 2GB limitation derives from OpenSSL <1.1.1, which used signed 32-bit ints ([bpo-42853 note](https://bugs.python.org/issue42853#msg384566)) in the relevant library function (`SSL_read`).

However, Python (cpython?) from 3.10 onwards calls the [more-recent](https://github.com/openssl/openssl/commit/4ee7d3f94590d3766c28eed03b969f338ddf96dc) [`SSL_read_ex`](https://github.com/python/cpython/blob/3.10/Modules/_ssl.c#L2513) function in OpenSSL that _does_ provide >2GB support.

So: an old Python compiled with OpenSSL >=1.1.1 would still have the limitation/bug, and Python with OpenSSL <1.1.1 would fail to compile in the first place (because the `SSL_read_ex` function wouldn't exist to link to).

Please double-check my logic there - but yep, I think I was mistaken, sorry for the distraction.

Comment (sethmlarson) [reply src/urllib3/response.py:715]: @jayaddison No problem, thanks for participating in the discovery. It's good to know the innards of the space we're working on! :)

File: src/urllib3/response.py
Comment (sethmlarson) [reply src/urllib3/response.py:721]: It looks like `BytesIO` is consistently faster than an unbounded `bytearray`?

```
$ python -m timeit -s 'import io; x = io.BytesIO()' 'for _ in range(10000):
    x.write(b"x" * 10000)

x.getvalue()
'
5 loops, best of 5: 45.9 msec per loop

$ python -m timeit -s 'import io; x = bytearray()' 'for _ in range(10000):
    x += b"x" * 10000

bytes(x)
'
2 loops, best of 5: 121 msec per loop
```

If that's the case we should use `io.BytesIO`, I wouldn't have guessed that!

File: src/urllib3/response.py
Comment (sethmlarson) [reply src/urllib3/response.py:732]: If the content is 3GB and the user calls `resp.read()` we have to load the whole 3GB into memory, there's no way around that unfortunately?

Comment (pquentin) [reply src/urllib3/response.py:732]: Sure but it's easy to accidentally use say 6GB instead.

Comment (sethmlarson) [reply src/urllib3/response.py:732]: Ah now I understand what you meant, wasn't thinking about overhead!

File: src/urllib3/response.py
Comment (illia-v) [reply src/urllib3/response.py]: Done in 890474a

File: src/urllib3/response.py
Comment (illia-v) [reply src/urllib3/response.py]: > Let's not consider unreleased versions of Python please :)

OK, I updated the condition in 49ed9a1

File: src/urllib3/response.py
Comment (illia-v) [reply src/urllib3/response.py]: I updated "cStringIO" to "StringIO" in 8fff25d, I was not sure if it is good to remove the comment

File: src/urllib3/response.py
Comment (illia-v) [reply src/urllib3/response.py]: BTW, if you know how to check whether a response is read via SSL, please let me know.

File: src/urllib3/response.py
Comment (pquentin) [reply src/urllib3/response.py]: Is this a good reason to use smaller chunks?

Comment (illia-v) [reply src/urllib3/response.py]: Good idea! The smaller the chunk, the smaller the overhead should be.

I would not make it too small though. Maybe 512 MiB?

Python used to read in 1 MiB-chunks. Obviously, there were complaints https://github.com/python/cpython/issues/80231.

**Comment by sethmlarson** [comment]
Huh, so SecureTransport only supports 16K reads at a time? Is that something wrong with our implementation or is that a ST quirk?

**Comment by illia-v** [comment]
`SSL_read` called by CPython and PyPy seems to return at most 16 KiB too. But when `http.client.HTTPResponse.read` is called with a bigger `amt`, they seem to combine the 16 KiB chunks so that the method returns the requested `amt` of bytes. `SecureTransport` seems to read only one 16 KiB chunk when `amt` is bigger than 16 KiB.

**Comment by illia-v** [comment]
PyPy tests are timing out because downloads with PyPy are very slow and memory-inefficient. One `test_requesting_large_resources_via_ssl` runs for 4.5 minutes locally, there are three variants in each of `test.with_dummyserver.test_socketlevel.TestSSL` and `test.contrib.test_pyopenssl.TestSocketSSL`, so the test is too slow for CI.

BTW, PyPy is prone to the overflow error too, the test results in `OverflowError: integer 2147483648 does not fit 'int'` in the main branch of urllib3. The change in this branch fixes the error and the tests pass.

**Comment by sethmlarson** [comment]
@illia-v Interesting! The SecureTransport issue seems like a separate issue and could impact the 1.26.x branch, could we try creating a separate PR which adds a test case that calls `resp.read(16K+1)` inside of `test/with_dummyserver/test_https.py` and asserts that 16K+1 bytes are returned for each TLS backend?

If we can get that issue fixed that would help us along with this PR as well.

File: src/urllib3/response.py
Comment (sethmlarson) [inline src/urllib3/response.py:686]: This statement is really complex and I don't like having the `version_info` patch version determining behavior, what are your thoughts on simplifying it to this (since only 3.7 and late 3.9 benefit from our escape hatch):

```python
        if (
            (
                (amt and amt > c_int_max)
                or (self.length_remaining and self.length_remaining > c_int_max)
            )
            and not util.IS_SECURETRANSPORT
            and (util.IS_PYOPENSSL or sys.version_info < (3, 10))
        ):
```

File: test/with_dummyserver/test_socketlevel.py
Comment (sethmlarson) [inline test/with_dummyserver/test_socketlevel.py:1505]: Can we try to manually run PyPy on a `sys.version_info` version less than 3.10 to see that our implementation isn't super slow like it is in the test suite? Would be good to have a comparison between using urllib3+PyPy on a response that's size 2**31-1 and 2**31

**Review by sethmlarson — COMMENTED** [review]
Left two comments, otherwise I think this PR is ready to ship :tada: Once we merge we can backport to 1.26.x

File: test/with_dummyserver/test_socketlevel.py
Comment (illia-v) [reply test/with_dummyserver/test_socketlevel.py:1505]: It took 5 minutes 27 seconds for PyPy 3.9 to run the three tests. When I changed `2**31` to `2**31 - 100` (this plus the headers length end ups being near the limit) in the test, it took 5 minutes and 16 seconds.

File: src/urllib3/response.py
Comment (illia-v) [reply src/urllib3/response.py:686]: I do not mind 🙂. I did this in 9a19fc0.

File: test/with_dummyserver/test_socketlevel.py
Comment (sethmlarson) [reply test/with_dummyserver/test_socketlevel.py:1505]: Great, thank you!

**Review by sethmlarson — APPROVED** [review]
This looks great, thanks @illia-v! Will merge once CI passes.

**Comment by sethmlarson** [comment]
Awesome, thanks @illia-v! Once you backport this feature+test to the 1.26.x branch you can collect the bounty.

**Pull request merged**
